File: iterations/tabletest/iteration-40/eval-27-convert-from-testng/outputs/src/test/java/com/example/ShippingCostCalculatorTest.java
package com.example;

import org.tabletest.junit.Description;
import org.tabletest.junit.TableTest;

import java.math.BigDecimal;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ShippingCostCalculatorTest {

    private static final List<Integer> DEFAULT_DIMENSIONS = List.of(10, 10, 10);

    private final ShippingCostCalculator calculator = new ShippingCostCalculator();

    @Description("""
        Dimensions are fixed at [10, 10, 10] (dimensional weight ~0.2kg), so the
        effective weight always equals the actual weight for these rows.
        Dimensional weight overriding actual weight is covered by
        appliesDimensionalWeightWhenGreaterThanActual. No fragile/insured/hazmat
        options are applied. Carrier is fixed to DHL since carrier does not
        affect price (see carrierDoesNotAffectPrice).
        """)
    @TableTest("""
        Scenario                | Region | Speed    | Weight | Base Rate?
        EU standard, light      | EU     | standard | 0.5    | 5.00
        EU standard, medium     | EU     | standard | 3.0    | 7.50
        EU standard, heavy      | EU     | standard | 10.0   | 12.50
        EU standard, very heavy | EU     | standard | 25.0   | 20.00
        EU express, light       | EU     | express  | 0.5    | 8.00
        EU express, medium      | EU     | express  | 3.0    | 12.00
        US standard, light      | US     | standard | 0.5    | 7.00
        US express, heavy       | US     | express  | 10.0   | 30.00
        """)
    void calculatesBaseRateByZoneAndWeight(String region, String speed, double weight, BigDecimal baseRate) {
        ShippingZone zone = new ShippingZone(region, speed);
        BigDecimal result = calculator.calculateShippingCost(
                zone, weight, DEFAULT_DIMENSIONS, new PackageOptions(), Carrier.DHL);
        assertCost(baseRate, result);
    }

    @Description("""
        Zone is fixed to EU standard. Actual weight is held at 1.0kg in both
        rows; only dimensions vary, isolating whether the billable weight comes
        from the actual weight or the dimensional (volumetric) weight.
        """)
    @TableTest("""
        Scenario                                | Dimensions   | Rate?
        Small package, actual weight used        | [10, 10, 10] | 5.00
        Large package, dimensional weight used    | [70, 50, 10] | 12.50
        """)
    void appliesDimensionalWeightWhenGreaterThanActual(List<Integer> dimensions, BigDecimal rate) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 1.0, dimensions, new PackageOptions(), Carrier.DHL);
        assertCost(rate, result);
    }

    @Description("""
        Zone is fixed to EU standard, weight fixed to 3.0kg (base rate 7.50),
        so any surcharge is due solely to the oversize dimension threshold (100).
        """)
    @TableTest("""
        Scenario              | Dimensions   | Total Cost?
        At the size limit     | [100, 5, 5]  | 7.50
        Just over the limit   | [101, 5, 5]  | 17.50
        """)
    void appliesOversizeSurchargeWhenAnyDimensionExceedsLimit(List<Integer> dimensions, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, dimensions, new PackageOptions(), Carrier.DHL);
        assertCost(totalCost, result);
    }

    @Description("""
        Zone is fixed to EU standard, weight fixed to 3.0kg (base rate 7.50).
        """)
    @TableTest("""
        Scenario                  | Fragile | Total Cost?
        Standard handling         | false   | 7.50
        Fragile surcharge applied | true    | 8.625
        """)
    void appliesFragileSurcharge(boolean fragile, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        PackageOptions options = new PackageOptions();
        options.setFragile(fragile);
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, DEFAULT_DIMENSIONS, options, Carrier.DHL);
        assertCost(totalCost, result);
    }

    @Description("""
        Zone is fixed to EU standard, weight fixed to 3.0kg (base rate 7.50).
        Premium is 0.6% of insured value, with a $3.00 minimum. Insured values
        of 200 and 500 both resolve to the $3.00 minimum (below and exactly at
        it); 1000 exceeds the minimum, so the computed premium is charged instead.
        """)
    @TableTest("""
        Scenario                             | Insured Value | Total Cost?
        No insurance                         |                | 7.50
        Below minimum premium, minimum charged | 200          | 10.50
        At minimum premium exactly            | 500            | 10.50
        Above minimum, premium charged        | 1000           | 13.50
        """)
    void appliesInsurancePremiumWithMinimum(BigDecimal insuredValue, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        PackageOptions options = new PackageOptions();
        options.setInsuredValue(insuredValue);
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, DEFAULT_DIMENSIONS, options, Carrier.DHL);
        assertCost(totalCost, result);
    }

    @Description("""
        Zone is fixed to EU standard, weight fixed to 3.0kg (base rate 7.50).
        """)
    @TableTest("""
        Scenario              | Handling | Total Cost?
        No special handling   |          | 7.50
        Hazmat handling       | hazmat   | 15.50
        """)
    void appliesHazmatHandlingFee(String handling, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        PackageOptions options = new PackageOptions();
        options.setHandling(handling);
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, DEFAULT_DIMENSIONS, options, Carrier.DHL);
        assertCost(totalCost, result);
    }

    @Description("""
        Zone is fixed to EU standard, weight fixed to 3.0kg (base rate 7.50).
        Verifies the order surcharges are combined in: hazmat/oversize fees are
        added to the base first, the fragile multiplier is applied next, and the
        insurance premium is added last (so it is not itself multiplied by the
        fragile rate).
        """)
    @TableTest("""
        Scenario             | Fragile | Handling | Insured Value | Total Cost?
        Fragile and insured  | true    |          | 200           | 11.625
        Fragile and hazmat   | true    | hazmat   |               | 17.825
        """)
    void combinesSurchargesInOrder(boolean fragile, String handling, BigDecimal insuredValue, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        PackageOptions options = new PackageOptions();
        options.setFragile(fragile);
        options.setHandling(handling);
        options.setInsuredValue(insuredValue);
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, DEFAULT_DIMENSIONS, options, Carrier.DHL);
        assertCost(totalCost, result);
    }

    @Description("""
        Zone is fixed to EU express, weight fixed to 3.0kg (base rate 12.00).
        The carrier parameter does not influence price for any combination of
        zone/weight/options; this table asserts that regardless of carrier.
        """)
    @TableTest("""
        Scenario                          | Carrier              | Total Cost?
        Same zone and weight, any carrier | {DHL, UPS, FEDEX}    | 12.00
        """)
    void carrierDoesNotAffectPrice(Carrier carrier, BigDecimal totalCost) {
        ShippingZone zone = new ShippingZone("EU", "express");
        BigDecimal result = calculator.calculateShippingCost(
                zone, 3.0, DEFAULT_DIMENSIONS, new PackageOptions(), carrier);
        assertCost(totalCost, result);
    }

    private static void assertCost(BigDecimal expected, BigDecimal actual) {
        assertEquals(0, expected.compareTo(actual), () -> "expected " + expected + " but was " + actual);
    }
}

File: iterations/tabletest/iteration-45/eval-25-convert-from-spock/outputs/src/test/kotlin/com/example/ShippingCostCalculatorTest.kt
package com.example

import org.junit.jupiter.api.Assertions.assertEquals
import org.tabletest.junit.Description
import org.tabletest.junit.TableTest
import java.math.BigDecimal

class ShippingCostCalculatorTest {

    private val calculator = ShippingCostCalculator()

    @Description("""
        Dimensions are held small (20 x 15 x 10) so dimensional weight never exceeds
        actual weight, and carrier is fixed to DHL — neither affects the base rate
        (see determinesEffectiveWeightFromDimensions and carrierDoesNotAffectCost).
        """)
    @TableTest("""
        Scenario                        | Region | Speed    | Weight | Base Rate?
        EU standard, light weight       | EU     | standard | 0.5    | 5.00
        EU standard, medium weight      | EU     | standard | 3.0    | 7.50
        EU standard, heavy weight       | EU     | standard | 10.0   | 12.50
        EU standard, very heavy weight  | EU     | standard | 25.0   | 20.00
        EU express, light weight        | EU     | express  | 0.5    | 8.00
        EU express, medium weight       | EU     | express  | 3.0    | 12.00
        US standard, light weight       | US     | standard | 0.5    | 7.00
        US express, heavy weight        | US     | express  | 10.0   | 30.00
        """)
    fun looksUpBaseRateByRegionSpeedAndWeight(region: String, speed: String, weight: Double, baseRate: BigDecimal) {
        val zone = ShippingZone(region, speed)
        val cost = calculator.calculateShippingCost(zone, weight, listOf(20, 15, 10), PackageOptions(), Carrier.DHL)
        assertEquals(0, baseRate.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, carrier is DHL, for all rows.
        """)
    @TableTest("""
        Scenario                                  | Weight | Length | Width | Height | Base Rate?
        Actual weight exceeds dimensional weight  | 3.0    | 30     | 20    | 15     | 7.50
        Dimensional weight exceeds actual weight  | 1.0    | 70     | 50    | 10     | 12.50
        """)
    fun determinesEffectiveWeightFromDimensions(weight: Double, length: Int, width: Int, height: Int, baseRate: BigDecimal) {
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, weight, listOf(length, width, height), PackageOptions(), Carrier.DHL)
        assertEquals(0, baseRate.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, weight is 3.0, carrier is DHL, for all rows.
        Oversize threshold is 100 (any dimension unit).
        """)
    @TableTest("""
        Scenario                                   | Length | Width | Height | Total Cost?
        Within the oversize threshold              | 30     | 20    | 15     | 7.50
        One dimension past the oversize threshold  | 120    | 5     | 5      | 17.50
        """)
    fun appliesOversizeFeePastThreshold(length: Int, width: Int, height: Int, totalCost: BigDecimal) {
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(length, width, height), PackageOptions(), Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, weight is 3.0, dimensions are 30 x 20 x 15, carrier is DHL, for all rows.
        """)
    @TableTest("""
        Scenario          | Fragile | Total Cost?
        Standard package  | false   | 7.50
        Fragile package   | true    | 8.625
        """)
    fun appliesFragileSurcharge(fragile: Boolean, totalCost: BigDecimal) {
        val options = PackageOptions(isFragile = fragile)
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), options, Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, weight is 3.0, dimensions are 30 x 20 x 15, carrier is DHL, for all rows.
        Insurance premium is a percentage of insured value, floored at a fixed minimum.
        """)
    @TableTest("""
        Scenario                              | Insured Value | Total Cost?
        Computed premium at or below minimum  | {200, 500}    | 10.50
        Computed premium above minimum        | 1000          | 13.50
        """)
    fun appliesInsurancePremiumWithMinimum(insuredValue: BigDecimal, totalCost: BigDecimal) {
        val options = PackageOptions(insuredValue = insuredValue)
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), options, Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, weight is 3.0, dimensions are 30 x 20 x 15, carrier is DHL, for all rows.
        """)
    @TableTest("""
        Scenario             | Handling | Total Cost?
        No special handling  |          | 7.50
        Hazmat handling      | hazmat   | 15.50
        """)
    fun appliesHazmatHandlingFee(handling: String?, totalCost: BigDecimal) {
        val options = PackageOptions(handling = handling)
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), options, Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is standard, weight is 3.0, dimensions are 30 x 20 x 15, carrier is DHL.
        Shows that the fragile multiplier applies to the base rate before the insurance
        premium is added: insurance is computed on, and added to, the already-multiplied
        base rather than being multiplied itself.
        """)
    @TableTest("""
        Scenario                     | Fragile | Insured Value | Total Cost?
        Fragile package, insured    | true    | 200           | 11.625
        """)
    fun appliesFragileMultiplierBeforeInsurancePremium(fragile: Boolean, insuredValue: BigDecimal, totalCost: BigDecimal) {
        val options = PackageOptions(isFragile = fragile, insuredValue = insuredValue)
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), options, Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Region is EU, speed is express, weight is 3.0, dimensions are 30 x 20 x 15.
        """)
    @TableTest("""
        Scenario                        | Carrier           | Total Cost?
        Cost is independent of carrier  | {DHL, UPS, FEDEX} | 12.00
        """)
    fun carrierDoesNotAffectCost(carrier: Carrier, totalCost: BigDecimal) {
        val zone = ShippingZone("EU", "express")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), PackageOptions(), carrier)
        assertEquals(0, totalCost.compareTo(cost))
    }
}
